src/combinatorics/hales_jewett.lean

```diff
@@ -0,0 +1,299 @@
+
+Note that this file does not use the same notation as Wikipedia.
+
+-/
+
+open_locale classical
+open_locale big_operators
+
+instance (n : pnat) : nonempty (fin n) := ⟨⟨0, n.pos⟩⟩
+
+/-- An induction principle for types analogous to `pnat.rec_on`. To prove a statement about
+all nonempty finite types, it suffices to consider `unit` and `option α` where we know the
+statement for `α`, provided that our statement is preserved by `≃`.-/
+lemma fintype_induction {P : Type → Prop} (of_equiv : ∀ {α β}, α ≃ β → P α → P β)
+  (h_unit : P unit) (h_option : ∀ {α} [fintype α] [nonempty α], P α → P (option α))
+  (α) [fintype α] [nonempty α] : P α :=
+begin
+  suffices : ∀ n : pnat, P (fin n),
+  { refine of_equiv (fintype.equiv_fin α).symm
+      (this ⟨fintype.card α, finset.card_pos.mpr finset.univ_nonempty⟩), },
+  intro n, apply n.rec_on,
+  exact of_equiv fin_one_equiv.symm h_unit,
+  exact λ _ ih, of_equiv (fin_succ_equiv _).symm (h_option ih),
+end
+
+namespace combinatorics
+
+/-- The type of combinatorial lines. A line `l : line m n` in the hypercube `n → m` defines a
+function `m → (n → m)` to the hypercube, such that for each coordinate `i : n`, the function
+`l x i` is either identically `x` or constant. We require lines to be nontrivial, in the sense
+that `l x i` is identically `x` for at least one `i`.
+
+Formally, a line is represented by the function `l.idx_fun : n → option m` which says whether
+`l x i` is identically `x` (corresponding to `l.idx_fun i = none`) or constantly `y`
+(corresponding to `l.idx_fun i = some y`). When `m` has size `1` there can be many elements of
+`line m n` defining the same function. -/
+structure line (m n : Type) : Type :=
+(idx_fun : n → option m)
+(proper : ∃ i, idx_fun i = none)
+
+namespace line
+
+/- This lets us treat a line `l : line m n` as a function `m → (n → m)`. -/
+instance (m n) : has_coe_to_fun (line m n) :=
+⟨λ _, m → n → m, λ l x i, (l.idx_fun i).get_or_else x⟩
+
+/-- The type of lines that are only one color except possibly at their endpoints. -/
+@[nolint has_inhabited_instance]
+structure almost_mono {m n k : Type} (C : (n → option m) → k) :=
+(line : line (option m) n)
+(color : k)
+(has_color : ∀ x : m, C (line (some x)) = color)
+
+/-- The type of families of lines such that
+- each line is only one color except possibly at its endpoint
+- the lines all have the same endpoint
+- the colors of the lines are distinct.
+Used in the proof `exists_mono_in_high_dimension`. -/
+structure color_focused {m n k} (C : (n → option m) → k) :=
+(lines : multiset (almost_mono C))
+(focus : n → option m)
+(is_focused : ∀ p : almost_mono C, p ∈ lines → p.line none = focus)
+(distinct_colors : (lines.map almost_mono.color).nodup)
+
+instance {m n k} (C : (n → option m) → k) : inhabited (color_focused C) :=
+⟨⟨0, λ _, none, λ _, false.elim, multiset.nodup_zero⟩⟩
+
+/-- The diagonal line. It is the identity at every coordinate. -/
+def diagonal (m n) [nonempty n] : line m n :=
```

Comment:
Perhaps this definition could do with lemmas describing it? I think I'd be happy with one saying how it behaves when you coerce to a function but if you can think of others feel free to add them!

Reply:
Sure, I added a `diagonal_apply` lemma.